Revise modular pipelines docs (#3948)
kedro-org/kedro: docs/source/nodes_and_pipelines/modular_pipelines.md

> # Pipeline reusage with namespaces

Reviewer: Maybe "Reuse pipelines using namespaces"? Reusage doesn't sound right :P

In many typical Kedro projects, a single (“main”) pipeline increases in complexity as the project evolves. To keep your project fit for purpose, we recommend you separate your code into different pipelines (modules) that are logically isolated and can be reused.

## How to reuse your pipelines

If you want to create a new pipeline that performs the same tasks as your existing pipeline (e.g., `data_science`), you can use the same `pipeline()` creation function as described in [previous manual](pipeline_introduction.md). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline creation code should look like this:

```python
def create_new_pipeline(**kwargs) -> Pipeline:
    return pipeline(
    [data_science], # A name of existing pipeline
    inputs = {"old_input__df_name" : "new_input_df_name"},
    outputs = {"old_output_df_name" : "new_output_df_name"},
    parameters = {"params: test_size1": "params: test_size2"},
    )
```
This means you can easily create multiple pipelines based on the `data_science` pipeline to test different approaches with various input datasets and model training parameters.

## What is a Namespace

> If you need to try different options for training your model, constantly overwriting your outputs for each new pipeline can become tedious. Each model's results (outputs) should be saved in new objects, which can be annoying to manage manually. Namespaces are a perfect solution for this. By setting a `namespace="namespace_name"` parameter for each new pipeline (based on an existing pipeline), you achieve automatic output and node isolation. This is done by adding a namespace_namenamespace_name.` prefix to each pipeline node and output dataset. Your pipeline creation code will look like this:

Reviewer: ```suggestion
If you need to try different options for training your model, constantly overwriting your outputs for each new pipeline can become tedious. Each model's results (outputs) should be saved in new objects, which can be annoying to manage manually. Namespaces are a perfect solution for this. By setting a `namespace="namespace_name"` parameter for each new pipeline (based on an existing pipeline), you achieve automatic output and node isolation. This is done by adding a `namespace_name` prefix to each pipeline node and output dataset. Your pipeline creation code will look like this:
```
Might also be worth showing the example of adding the namespace prefix in the catalog and parameter files?

```python
def create_new_pipeline(**kwargs) -> Pipeline:
    return pipeline(
    [data_science], # A name of existing pipeline
    inputs = {"old_input_df_name" : "new_input_df_name"},
    parameters = {"params: test_size1": "params: test_size2"},
    namespace = "alternative_approach",
    )
```
* You can use `kedro run --namespace = namespace_name` to run only the specific namespace
* [Kedro-Viz](https://demo.kedro.org) accelerates development by rendering namespaced pipelines as collapsible 'super nodes'.

## How to share your pipelines

> Pipelines are shareable between Kedro codebases via [micro-packaging](micro_packaging.md), but you must follow a couple of rules to ensure portability:

Reviewer: Micropackaging is going to be deprecated so perhaps it's worth adding alternate instructions on how to do this?

> # Reuse pipelines with namespaces

Reviewer 2: This `modular_pipelines.md` could be `namespaces.md`, and then `modular_pipelines.md` should be for the last "how-to" sections of `pipeline_introduction.md` (see review comment)

> In many typical Kedro projects, a single (“main”) pipeline increases in complexity as the project evolves. To keep your project fit for purpose, we recommend you separate your code into different pipelines (modules) that are logically isolated and can be reused.

Reviewer 3: I don't think this is the purpose of namespace. Even without namespace, users can have multiple pipeline. i.e.

- pipelines
  - data_science
  - data_processing

## How to reuse your pipelines

> If you want to create a new pipeline that performs the same tasks as your existing pipeline (e.g., `data_science`), you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](pipeline_introduction.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline creation code should look like this:

Reviewer 3: ```suggestion
If you want to create a new pipeline that performs the same tasks as your existing pipeline (e.g., `data_science`), you can use the same `pipeline()` creation function as described in [How to structure your pipeline creation](pipeline_introduction.md#how-to-structure-your-pipeline-creation). This function allows you to overwrite inputs, outputs, and parameters. Your new pipeline creation code should look like this:
```

Similar tasks with different inputs/outputs?

```python
def create_new_pipeline(**kwargs) -> Pipeline:
    return pipeline(
    [data_science], # Name of an existing pipeline

>     inputs = {"old_input__df_name" : "new_input_df_name"},

Reviewer 3: ```suggestion
    inputs = {"old_input_df_name" : "new_input_df_name"},
```

* We recommend that you document the configuration required (parameters and catalog) in the local `README.md` file for any downstream consumers.

### Providing modular pipeline specific dependencies
## Providing pipeline specific dependencies

* A modular pipeline **might** have external dependencies specified in a local `requirements.txt` file.
* A pipeline **might** have external dependencies specified in a local `requirements.txt` file.
* Pipeline specific dependencies are scooped up during the [micro-packaging](micro_packaging.md) process.
* These dependencies are _not_ currently installed by the [`kedro install`](../development/commands_reference.md#install-all-package-dependencies) command, and must be manually installed.

## Using the modular `pipeline()` wrapper to provide overrides

This wrapper really unlocks the power of modular pipelines.

* It allows you to start instantiating the same pipeline multiple times.
* These will be static in terms of structure, but dynamic in terms of `inputs`/`outputs`/`parameters`.
* It also allows you to simplify both your mental models, and Kedro-Viz visualisations via `namespaces`.

```python

> from kedro.pipeline.modular_pipeline import pipeline

Reviewer 2: Happy to see this line go away 😄

# Modular pipelines

In many typical Kedro projects, a single (“main”) pipeline increases in complexity as the project evolves. To keep your project fit for purpose, we recommend you **separate your code into different pipelines (modules)** that are logically isolated and can be reused. Each pipeline should ideally be organised in its own folder, promoting easy copying and reuse within and between projects. Simply put: one pipeline, one folder.

Kedro supports this concept of modular pipelines with the following tools:
- [How to create a new blank pipeline using the `kedro pipeline create` command](#how-to-create-a-new-blank-pipeline-using-the-kedro-pipeline-create-command)
- [How to structure your pipeline creation](#how-to-structure-your-pipeline-creation)
- [How to use custom new pipeline templates](#how-to-use-custom-new-pipeline-templates)
- [How to share your pipelines](#how-to-share-your-pipelines)

> ## How to create a new blank pipeline using the `kedro pipeline create` command

Reviewer 4: Long subheads like this look very awkward in the left hand table of contents. Is it necessary to have the full title or could it just be "How to create a new blank pipeline" ?

> To assist with creating a new pipeline that supports the modularity concept, you can use the following command (the pipeline name must adhere to [Python naming conventions](https://realpython.com/python-pep8/#naming-conventions)):

Reviewer 4: > To assist with creating a new pipeline that supports the modularity concept, you can use the following command (the pipeline name must adhere to [Python naming conventions](https://realpython.com/python-pep8/#naming-conventions)):

This is a verbose way to basically just say:
```suggestion
 To create a new modular pipeline, use the following:
```

    │   ├── __init__.py
    │   └── pipelines
    │       ├── __init__.py

>     │       └── {{pipeline_name}}      <-- This folder defines the pipeline

Reviewer 2: Maybe keep "modular" here

```

If you want to delete an existing pipeline, you can use `kedro pipeline delete <pipeline_name>` to do so.
```{note}

> To see the full list of available CLI options, you can run `kedro pipeline create --help` for more information.

Reviewer 2: ```suggestion
To see the full list of available CLI options, you can run `kedro pipeline create --help`.
```

?

```

## How to structure your pipeline creation

After creating the pipeline with `kedro pipeline create`, you will find template code in `pipeline.py` that you need to fill with your actual pipeline code:

```python
# src/my_project/pipelines/{{pipeline_name}}/pipeline.py
from kedro.pipeline import Pipeline, pipeline

def create_pipeline(**kwargs) -> Pipeline:
    return pipeline([])
```
Here, you are creating a function that returns a `Pipeline` class instance with the help of the `pipeline` function. We recommend putting your pipeline creation code into a function because it allows to achieve:

> - Modularity: you can easily reuse this logic across different parts of your project.

Reviewer 2: I'd maybe skip this bullet list, or I'd move it to the top of the page, where the concept is explained and justified

Author: I changed them with explanation that function should be named `create_pipeline()` for auto-discovery.

Kedro supports having a single pipeline template in your project. If you need to have multiple pipeline templates, consider saving them in a
separate folder and pointing to them with the `--template` flag.

> ### Creating custom templates

Reviewer 2: ```suggestion
### Creating custom pipeline templates
```
